packages/expo-modules-core/android/src/main/java/expo/modules/core/logging/PersistentFileLog.kt

```diff
@@ -0,0 +1,143 @@
+package expo.modules.core.logging
+
+import android.content.Context
+import java.io.File
+import java.io.IOException
+import java.nio.charset.Charset
+
+/**
+ * A thread-safe class for reading and writing line-separated strings to a flat file
+ * The main use case is for logging specific errors or events, and ensuring that the logs
+ * persist across application crashes and restarts (for example, logcat can only read system logs
+ * for the current process, and cannot access anything logged before the current process started).
+ *
+ * All write access to the file goes through asynchronous public methods managed by
+ * a static serial dispatch queue implemented with Kotlin coroutines. This ensures that
+ * multiple instances accessing the file will have thread-safe access.
+ *
+ * The only operations supported are
+ * - Read the file (synchronous)
+ * - Append one or more entries to the file
+ * - Filter the file (only retain entries that pass the filter check)
+ * - Clear the file (remove all entries)
+ *
+ */
+class PersistentFileLog(
+  category: String,
+  context: Context
+) {
+
+  /**
+   * Read entries from log file
+   */
+  fun readEntries(): List<String> {
+    if (0L == getFileSize()) {
+      return listOf()
+    }
+    return readFileLinesSync()
+  }
+
+  /**
+   * Append entry to the log file
+   * Since logging may not require a result handler, the handler parameter is optional
+   */
+  fun appendEntry(entry: String, completionHandler: ((_: Error?) -> Unit) = { }) {
+    queue.add {
+      try {
+        this.ensureFileExists()
+        val text = when (this.getFileSize()) {
+          0L -> entry
+          else -> {
+            "\n" + entry
+          }
+        }
+        this.appendTextToFile(text)
+        completionHandler.invoke(null)
```

Comment:
Sorry I didn't notice this earlier, but I think it's worth looking at. Looking at the try/catch here (and in all the other methods in this file), this will catch errors thrown in the callback, and then will call the completion handler twice if the callback throws. I think there's a couple things that might be causes for concern:
- I'm not sure whether this library should be responsible for catching errors in the callback. I don't really have alternate suggestions. It very well might be correct as-is, in which case we need to decide...
- Whether calling the callback twice makes sense in this case.

Since this is a first party module and only used so far in first party modules, it probably is fine to leave it as-is for now since we don't expect our callbacks to throw. But, if we open this up in the future, we may want to consider a different error pattern.

For now though, this is fine as-is I think. Might just be worth filing a linear task to think about this in the future.

Reply:
@wschurman I'm now running the completion handler on the main queue, which should solve this issue.